demo_nodes_cpp/src/parameters/parameter_events.cpp

```diff
@@ -66,6 +66,12 @@ int main(int argc, char ** argv)
       on_parameter_event(event, node->get_logger());
     });
 
+  // Declare parameters that may be set on this node
+  node->declare_parameter("foo");
+  node->declare_parameter("bar");
+  node->declare_parameter("baz");
+  node->declare_parameter("foobar");
```

Comment:
Was this test failing? Why are these needed here now?

Reply:
The original template expected these parameters to show under `new parameters:`. That wasn't the case when these declarations were before an `on_parameters_set` callback was set.

Reply:
As to why these didn't fail before, I think during the migration I changed output matching to be a bit more strict (like, order of appearance has to be respected now).

Reply:
This isn't a completely satisfactory answer for me, since this node is not created with the `allow_undeclared_parameters` option explicitly set to `true` (default is `false`), so either the declare parameters would have been always required or are not required now. I just don't understand why this is now required.

I would say that you should either avoid the change in output matching that you theorize makes this necessary, or refactor this test completely (including explaining why this is needed and fixing the timing issue), but the current in between patch state is kind of the worst of both. We don't know why we're changing the test, if it was invalid before, or made invalid after these changes, and we've made it more flaky, the resolution of which is to just wait longer, not making reliable, just making it less likely to flake.

Reply:
I'm not adding parameter declarations here, just moving them around. What I've just realized is that, in this PR, that's not observable because declarations were added by https://github.com/ros2/demos/pull/241, [here](https://github.com/ros2/demos/pull/241/files#diff-d205843bf5f21190ea0e0c9955de833dR31), and don't show up until you merge this branch into `master`. Which means this needs a rebase or we'll end up with the same declarations twice.

As to why it didn't fail before, I'm not theorizing. Compare the [new output matching code](https://github.com/ros2/launch/pull/215/files#diff-fc18b5a32be704c451cc2722e73d6e57R26) with the [current output matching code](https://github.com/ros2/launch/blob/master/launch_testing/launch_testing/output.py#L100). The second will match the pattern anywhere in the output, while the first will prune the output as it matches to enforce an order.

I can write a small repro tomorrow if you're not sufficiently convinced though.

Reply:
> I'm not adding parameter declarations here, just moving them around.

I don't see that in the diff...

> What I've just realized is that, in this PR, that's not observable because declarations were added by #241, here, and don't show up until you merge this branch into master. Which means this needs a rebase or we'll end up with the same declarations twice.

Well that's the missing piece. Without the rebase you're duplicating what's on master in order to get it to pass. Which is still odd because unless you're using a custom repos file, our CI should be merging with master already...

> As to why it didn't fail before, I'm not theorizing. Compare the new output matching code with the current output matching code. The second will match the pattern anywhere in the output, while the first will prune the output as it matches to enforce an order.

I'm not concerned about whether or not it's theory or proven, but that we're changing this test to accommodate changes in the launch system. Put another way, why did you decide to change this test rather than change the output matching so this test passed without modification? Was the flexibility of the previous output matching a limitation of the approach or a feature?

With the understanding that the declare parameters are on master, and missing from this pr, I now understand why the test was passing before, but I still don't know if reordering them here is the right thing to do (or even intentional). If you think it is what we should do, that's fine and I'll go along with it, but it was not clear from the diff nor pr descriptions why this was needed at all.

Reply:
Now that you've rebased, why does this block need to be moved down? I had assumed you not only moved the declare's but changed their order, but it looks like the order is the same.

Reply:
> Without the rebase you're duplicating what's on master in order to get it to pass.

Precisely.

> but that we're changing this test to accommodate changes in the launch system. 

Allow me to disagree here. We're not changing something to get it to pass a launch test, but rather to fix it, since the previous test was broken due to how output matching was performed and where parameters were declared on #241. Let me explain a bit further.

The current, soon to be deprecated output matching would attempt to match each output line separately against the entire collection of expected lines. As a result, this output:

```
Parameter event:
 new parameters:
 changed parameters:
  foobar
 deleted parameters:
```

and this output:

```
Parameter event:
 new parameters:
  foobar
 changed parameters:
 deleted parameters:
```

both match this pattern:

```
Parameter event:
 new parameters:
  foobar
 changed parameters:
 deleted parameters:
```

That's wrong but nonetheless the behavior of the old legacy output matching code and the one that was migrated back then.

The original pattern for parameter event tests was there to match the output that'd result of an initial parameter set (an implicit declaration before #241) and a follow up modification. After #241, by putting the declaration block before the callback setting, we stopped getting output for declarations but because of the defective output matching CI didn't say squat.

Now that output matching respects ordering, the output no longer matched the pattern and it started to fail. Thus, I moved the entire declaration block as-is after callback setting to get output as a result of declarations. That made it pass, but one could argue that it's insufficient as we're not testing all blocks, declaration, first and second modification, separately. I've modified the output pattern in https://github.com/ros2/demos/pull/318/commits/798105a652673e4309526e07ee536fd02fe07b07 to validate the output of each block.

Hopefully the change is now more clear.

Reply:
Ok, sounds good, thanks for explaining.